### DISCUSSION TITLE
Params should be configurable in next/image cloudinaryLoader

### DISCUSSION THREAD

**Discussion by neeraj3029** [vercel/next.js#35017] [Ideas]
### Describe the feature you'd like to request

**The user should be able to dictate cloudinary params as per their requirements. If not provided, we can fallback to the default params.**

Currently, the user is only able to add `width` and `quality` as params to the `cloudinaryLoader`. For my use case, the cloudinary link should consist of other params for image size, for eg `t_web_ab_2x`. 

For example, People might not use `width` as attribute to specify image dimensions, they could other custom ways like `t_web_ab_4x` etc.

**Source Code reference:** https://github.com/vercel/next.js/blob/ed43c03ebb164d88657c933be569b499f8e1e074/packages/next/client/image.tsx#L813

```
// Current Implementation
function cloudinaryLoader({
  config,
  src,
  width,
  quality,
}: ImageLoaderProps): string {
  // Demo: https://res.cloudinary.com/demo/image/upload/w_300,c_limit,q_auto/turtles.jpg
  const params = ['f_auto', 'c_limit', 'w_' + width, 'q_' + (quality || 'auto')]
  const paramsString = params.join(',') + '/'
  return `${config.path}${paramsString}${normalizeSrc(src)}`
}
```

### Describe the solution you'd like

The params should be passed as argument to the cloudinaryLoader. We can have some default params if the user hasn't provided any.

Ideal implementation as per me:

```
function cloudinaryLoader({
  config,
  src,
  width,
  quality,
  params
}: ImageLoaderProps): string {
  // Demo: https://res.cloudinary.com/demo/image/upload/w_300,c_limit,q_auto/turtles.jpg
  const cloudinaryParams = params || ['f_auto', 'c_limit', 'w_' + width, 'q_' + (quality || 'auto')]
  const paramsString = cloudinaryParams.join(',') + '/'
  return `${config.path}${paramsString}${normalizeSrc(src)}`
}
```

where
```
params: ['f_auto', 't_web_ab_2x']
```

**Comment by neeraj3029** [comment]
@styfle @ijjk PTAL

**Comment by styfle** [comment]
You can achieve this behavior today using the [`loader`](https://nextjs.org/docs/api-reference/next/image#loader) prop

**Reply by chrillep** [reply]
> You can achieve this behavior today using the [`loader`](https://nextjs.org/docs/api-reference/next/image#loader) prop

@styfle could you give a code example on how to pass more props (params) to a custom loader (cloudinary) in typescript?

**Reply by styfle** [reply]
The link above has an example

```js
import Image from 'next/image'

const myLoader = ({ src, width, quality }: {src: string, width: number, quality?: number}) => {
  return `https://example.com/${src}?w=${width}&q=${quality || 75}`
}

return <Image
  loader={myLoader}
  src="me.png"
  alt="Picture of the author"
  width={500}
  height={500}
/>
```

**Reply by chrillep** [reply]
> The link above has an example
> 
> ```js
> import Image from 'next/image'
> 
> const myLoader = ({ src, width, quality }: {src: string, width: number, quality?: number}) => {
>   return `https://example.com/${src}?w=${width}&q=${quality || 75}`
> }
> 
> return <Image
>   loader={myLoader}
>   src="me.png"
>   alt="Picture of the author"
>   width={500}
>   height={500}
> />
> ```
@styfle sry but i don't get it 😓  . this doesn't pass any extra params to the loader?

for example adding aspectRatio
could be '1:1' , '3:4' or '5:6'

```tsx
import Image from 'next/image';

const myLoader = ({
  src,
  width,
  quality,
  aspectRatio,
}: {
  src: string;
  width: number;
  quality?: number;
  aspectRatio: '1:1' | '3:4' | '5:6';
}) => {
  return `https://example.com/${src}?ar=${aspectRatio}?w=${width}&q=${quality || 75}`;
};

return (
  <Image
    loader={myLoader}
    src="me.png"
    alt="Picture of the author"
    width={500}
    height={500}
  />
);
```
how would one pass aspectRatio to the loader via next/image

**Reply by chrillep** [reply]
you could create three different loaders, one for each aspectRatio is that what u mean ?

**Reply by styfle** [reply]
Sure you could have 3 different loaders or create 1 function with a closure that returns another function.

```js
function getLoader(aspectRatio) {
  const myLoader = ({
    src,
    width,
    quality,
    aspectRatio,
  }: {
    src: string;
    width: number;
    quality?: number;
    aspectRatio: '1:1' | '3:4' | '5:6';
  }) => {
    return `https://example.com/${src}?ar=${aspectRatio}?w=${width}&q=${quality || 75}&aspectRatio=${aspectRatio}`;
  };
  return myLoader;
}
```

**Reply by chrillep** [reply]
> Sure you could have 3 different loaders or create 1 function with a closure that returns another function.

I'll give it a try! 😊
will the typings work for that in ts?

```tsx
import Image, { ImageLoader } from 'next/image';

function getLoader(aspectRatio: '1:1' | '3:4' | '5:6'): ImageLoader {
  const myLoader = ({
    src,
    width,
    quality,
  }: {
    src: string;
    width: number;
    quality?: number;
  }) => {
    return `https://example.com/${src}?ar=${aspectRatio}?w=${width}&q=${quality || 75}`;
  };
  return myLoader;
}
```
worked!
ty! cheers!

**Comment by chrillep** [comment]
@neeraj3029 found this btw
https://www.npmjs.com/package/next-image-cloudinary

https://github.com/kedarguy/next-image-cloudinary/blob/d769c40e4ef2586344718a51edeebd8a7cf84c9c/src/index.tsx#L1-L31

it uses https://www.npmjs.com/package/cloudinary-build-url
